File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B02_kmedoides_gower/R/B02_kmedoides_gower.R
# ############################################################################
# B2. K-MEDOIDES
# Tema B02_kmedoides_gower — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B02_kmedoides_gower.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B2. K-MEDOIDES\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B2. K-MEDOIDES (PAM) Y DISTANCIA DE GOWER
# ----------------------------------------------------------------------------
# QUÉ HACE: igual que K-Means, pero el centro de cada grupo es UNA
# OBSERVACIÓN REAL (el medoide), no un promedio inventado.
#
#   medoide_k = la observación de C_k que minimiza la suma de distancias
#               a las demás del grupo
#
# POR QUÉ IMPORTA
#   1. ROBUSTO a outliers: un punto extremo arrastra un promedio, pero no
#      puede "sacar" al medoide de los datos.
#   2. Solo necesita una MATRIZ DE DISTANCIAS => sirve con variables
#      CATEGÓRICAS o MIXTAS usando GOWER.
#
# GOWER, EN SIMPLE: compara dos observaciones variable por variable. Si es
# numérica, la diferencia escalada por el rango; si es categórica, 0 si son
# iguales y 1 si no. Luego PROMEDIA. Da un número entre 0 y 1.
#
# PARÁMETROS de pam()
#   k     : número de grupos
#   diss  : TRUE si el primer argumento ya es una matriz de distancias
#   La DECISIÓN MÁS IMPORTANTE es la distancia, no k.
#
# ❌ COSTOSO: la matriz es n x n. Con n > 10.000 ya pesa -> usa clara()
# ============================================================================

library(cluster)
K <- 3

# Opción A: distancia euclidiana (variables numéricas)
D <- dist(Xs)

# Opción B: distancia de GOWER (variables mixtas / categóricas)
#   D <- daisy(X_df, metric="gower")

pam_res <- pam(D, k=K, diss=TRUE)

cat("medoides (filas del dataset):", pam_res$id.med, "\n")
cat("tamaño de cada grupo:", table(pam_res$clustering), "\n")
cat("\nLos medoides son observaciones REALES:\n")
print(round(X_df[pam_res$id.med, ], 2))

plot(Z2, col=pam_res$clustering, pch=16, xlab="PC1", ylab="PC2",
     main="K-Medoides")
points(Z2[pam_res$id.med, , drop=FALSE], col="red", pch=4, cex=3, lwd=4)
legend("topright", "medoides", col="red", pch=4)

# Silueta (funciona con CUALQUIER distancia)
sil <- silhouette(pam_res)
plot(sil, main="Silueta - K-Medoides")
cat(sprintf("ancho promedio de silueta: %.3f\n", mean(sil[,"sil_width"])))

# Para n grande: clara(X_df, k=3, samples=50)

cat("\nListo. Revisa la carpeta figuras/\n")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B02_kmedoides_gower/python/B02_kmedoides_gower.py
"""
B2. K-Medoides (PAM) y distancia de Gower
Tema B02_kmedoides_gower — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B02_kmedoides_gower.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B2. K-Medoides (PAM) y distancia de Gower")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B2. K-Medoides ═══
import kmedoids
from scipy.spatial.distance import squareform, pdist

K = 3

# Opción A: distancia euclidiana (variables numéricas)
D = squareform(pdist(Xs, metric="euclidean"))

# Opción B: distancia de Gower (variables mixtas / categóricas)
#   import gower
#   D = gower.gower_matrix(X_df.astype(object)).astype(np.float64)

res = kmedoids.fasterpam(D, medoids=K)
print(f"pérdida total : {res.loss:.3f}")
print(f"medoides      : filas {res.medoids} del dataset")
print(f"tamaño grupos : {np.bincount(res.labels)}")
print("\nLos medoides son observaciones REALES:")
print(X_df.iloc[res.medoids].round(2).to_string())

plt.scatter(Z2[:,0], Z2[:,1], c=res.labels, cmap="viridis", alpha=.7)
plt.scatter(Z2[res.medoids,0], Z2[res.medoids,1], c="red", marker="X",
            s=250, edgecolors="black", label="medoides")
plt.xlabel("PC1"); plt.ylabel("PC2"); plt.title("K-Medoides"); plt.legend(); plt.show()

print("\nListo. Revisa la carpeta figuras/")
